src/source-af-xdp.c

```diff
@@ -0,0 +1,787 @@
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    SCLogDebug("bind to %s on queue %u", ptv->iface, ptv->xsk.queue_num);
+
+    /* For polling and socket options */
+    ptv->xsk.fd.fd = xsk_socket__fd(ptv->xsk.xsk);
+    ptv->xsk.fd.events = POLLIN;
+
+    SCMutexUnlock(&xsk_protect.queue_protect);
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static inline int DumpStatsEverySecond(AFXDPThreadVars *ptv, time_t *last_dump)
+{
+    int stats_dumped = 0;
+    time_t current_time = time(NULL);
+
+    if (current_time != *last_dump) {
+        AFXDPDumpCounters(ptv);
+        *last_dump = current_time;
+        stats_dumped = 1;
+    }
+
+    StatsSyncCountersIfSignalled(ptv->tv);
+
+    return stats_dumped;
+}
+
+/**
+ *  \brief Main AF_XDP reading Loop function
+ */
+static TmEcode ReceiveAFXDPLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    Packet *p;
+    time_t last_dump = 0;
+    uint32_t idx_rx = 0, idx_fq = 0, rcvd;
+    AFXDPThreadVars *ptv = (AFXDPThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    /* Reserve memory for umem  */
+    if (ConfigureXSKUmem(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    /* Open AF_XDP socket */
+    if (OpenXSKSocket(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (ConfigureBusyPolling(ptv) != TM_ECODE_OK) {
+        SCLogWarning(SC_WARN_AFXDP_CONF, "Failed to configure busy polling"
+                                         " performance may be reduced.");
+    }
+
+    /* Attach eBPF program to interface */
+    if (bpf_get_link_xdp_id(ptv->ifindex, &ptv->xsk.prog_id, ptv->xsk.cfg.xdp_flags)) {
+        SCLogError(SC_ERR_BPF, "Failed to attach eBPF program to interface: %s", ptv->livedev->dev);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (InitFillRing(ptv, NUM_FRAMES * 2) != TM_ECODE_OK) {
+        SCLogError(SC_ERR_AFXDP_INIT, "Failed to initialise the fill ring.");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    AFXDPAllThreadsRunning(ptv);
+
+    PacketPoolWait();
+    while (1) {
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogDebug("Stopping Suricata!");
+            AFXDPDumpCounters(ptv);
+            break;
+        }
+
+        /* Busy polling is not set, using poll() to maintain (relatively) decent
+         * performance. xdp_busy_poll must be disabled for kernels < 5.11
+         */
+        if (!ptv->xsk.enable_busy_poll) {
+            StatsIncr(ptv->tv, ptv->capture_afxdp_poll);
+
+            int ret = poll(&ptv->xsk.fd, 1, POLL_TIMEOUT);
+
+            /* Report poll results */
+            if (ret <= 0) {
+                if (ret == 0) {
+                    StatsIncr(ptv->tv, ptv->capture_afxdp_poll_timeout);
+                } else if (ret < 0) {
+                    StatsIncr(ptv->tv, ptv->capture_afxdp_poll_failed);
+                    SCLogWarning(SC_ERR_AFXDP_READ, "poll failed with retval %d", ret);
```

Comment:
I conducted a test to ensure if the interface dropped and recovered the application did not become unresponsive and it continued to receive packets after interface recovery. 

My test method for this was to have suricata running and receiving packets, switching the interface to DOWN and confirming packets were not received, switching the interface to UP and confirming it started to receive packets. 

I confirmed this with: 
- Ethernet controller: VMware VMXNET3 Ethernet Controller, driver: vmxnet3
- Ethernet controller: Intel Corporation Ethernet Controller XL710 for 40GbE QSFP+, driver: i40e

Reply:
I have implemented a retry/reopen mechanism, similar to ``source-af-packet.c``. Upon detection of ``poll() | recv()`` failure the state will be changed to ``AFXDP_STATE_DOWN`` and the system will wait for the interface to recover. It will log a warning and information of these events, respectfully. This will be introduced in the next PR.